### ISSUE TITLE
remove slot# from account's hash

### ISSUE THREAD

**Issue by jeffwashington** [solana-labs/solana#28404]
#### Problem

tl;dr:
Including the slot # in an account's hash is unnecessary and inefficient.
Let's remove the slot # from an account's hash calculation (using a feature).
prototype: #28405

Details:

It is inefficient to keep track of an account's hash. It increases appendvec sizes (and thus snapshot size and ramdisk size) and reduces compress-ability. The hash can be recalculated today, given the slot #.

In the near future, we will stop doing rewrites and we will begin accumulating a large, sparse set of old slots with a few accounts each.

It will be inefficient to keep track of the slot that an account was last written to once the slot becomes ancient (> 432k slots) and is expected to contain few accounts.

An observation is that including the slot # in an account's hash does not provide any additional security, protection, or confirmation that an account is 'correct' in the system.

We use the account's hash in two algorithms.

1. bank hash, which includes the accounts delta hash
accounts delta hash is computed by 
a. getting all accounts store in the bank's slot
b. sorting by pubkey
c. hashing the account hash for each account

Today, each account's hash includes the slot. They will and must all have the same slot, the same slot as the bank itself. This provides no useful information and is just redundant.

2. accounts hash
accounts hash is calculated by (conceptually):
a. iterating over all append vecs <= a given slot, with append vecs ordered by slot from low to high)
b. iterating over each account in each append vec
c. keeping track of the latest (pubkey, hash) tuple by pubkey. A (pubkey, hash) found later in the same append vec or in a higher slot with the same pubkey overwrites the previous entry for that pubkey).
d. sorting by pubkey, then hashing the hashes in a merkle tree

An accounts hash verifies that the entire state of all accounts is consistent with the network.

So, what does it mean to verify that the entire state of all accounts is consistent with the network?
On account load, the accounts db is asked to return the most recently written account data of a pubkey, given ancestors and roots. No caller[1][2] cares which SPECIFIC slot the account was loaded from. The caller always only cares about it being the most recently written version of the account.
[1] internal operations like clean and shrink don't care about specific slot #s either except to compare 2 slots relative to each other and intersect them with roots, ancestors, and such. An account will still be associated internally with a slot, but that slot just has to be in the correct order relative to other versions of the same pubkey.
[2] RPC clients may specify a commitment, which ends up being a maximum slot that accounts are loaded up to. The specific slot that each account was last modified is not important.

From this, we see:
1. the state of the system is 'correct' if for any pubkey, loading the account data for that pubkey returns the most recent, correct account data.
2. the specific slot # associated with the most recent account data is irrelevant
3. whether an account's hash includes the slot # of the most recent store of that account or not has no impact on the correctness of the system.
4. Since the accounts hash calculation is designed to verify the state of the accounts is correct, it does not matter whether the account's hash includes the slot # or not - as long as the most recent account data is what is hashed into the account's hash.

Conclusion:

So, to improve performance of the system and simplify other aspects going forward, I propose to eliminate the slot # from the account's hash calculation.
#28405

**Comment by jeffwashington** [comment]
> any concerns with data integrity in a potential future where we resume from on disk accounts state rather than a snapshot?

This will work as expected.

We will still maintain and verify that we will correctly answer the question: what is the most recent account data for any pubkey as of slot X, if slot X = the restoration slot.
